packages/material-ui/src/Portal/Portal.js

```diff
@@ -17,65 +18,64 @@ function getOwnerDocument(element) {
  * Portals provide a first-class way to render children into a DOM node
  * that exists outside the DOM hierarchy of the parent component.
  */
-class Portal extends React.Component {
-  componentDidMount() {
-    this.setMountNode(this.props.container);
+function Portal(props) {
+  const { children, disablePortal, onRendered, container } = props;
+  const [, forceUpdate] = React.useState(0);
+  const [childNode, setChildNode] = React.useState();
+  const mountNode = React.useRef();
+  const renderedTimer = React.useRef();
+
+  const setMountNode = React.useCallback(() => {
+    if (disablePortal) {
+      mountNode.current = ReactDOM.findDOMNode(childNode).parentElement;
 
-    // Only rerender if needed
-    if (!this.props.disablePortal) {
-      this.forceUpdate(this.props.onRendered);
+      return;
     }
-  }
-
-  componentDidUpdate(prevProps) {
-    if (
-      prevProps.container !== this.props.container ||
-      prevProps.disablePortal !== this.props.disablePortal
-    ) {
-      this.setMountNode(this.props.container);
+    mountNode.current = getContainer(container, getOwnerDocument(childNode).body);
+  }, [container, disablePortal]);
 
-      // Only rerender if needed
-      if (!this.props.disablePortal) {
-        this.forceUpdate(() => {
-          if (this.props.onRendered) {
-            // This might be triggered earlier than the componentDidUpdate of a parent element.
-            // We need to account for it.
-            clearTimeout(this.renderedTimer);
-            this.renderedTimer = setTimeout(this.props.onRendered);
-          }
-        });
-      }
-    }
-  }
+  const handleOwnRef = React.useCallback(ref => {
+    setChildNode(ReactDOM.findDOMNode(ref));
+  }, []);
+  const handleRef = useForkRef(children.ref, handleOwnRef);
+  const child = React.cloneElement(children, {
+    ref: handleRef,
+    ...children.props,
+  });
 
-  componentWillUnmount() {
-    this.mountNode = null;
-    clearTimeout(this.renderedTimer);
-  }
+  React.useEffect(() => {
+    setMountNode();
 
-  setMountNode(container) {
-    if (this.props.disablePortal) {
-      this.mountNode = ReactDOM.findDOMNode(this).parentElement;
-      return;
+    // Only rerender if needed
+    if (!disablePortal) {
+      forceUpdate(onRendered);
     }
 
-    this.mountNode = getContainer(container, getOwnerDocument(this).body);
-  }
-
-  /**
-   * @public
-   */
-  getMountNode = () => this.mountNode;
-
-  render() {
-    const { children, disablePortal } = this.props;
+    return () => {
+      mountNode.current = undefined;
+      clearTimeout(renderedTimer.current);
+    };
+  }, [container, disablePortal, onRendered, setMountNode]);
 
-    if (disablePortal) {
-      return children;
+  React.useEffect(() => {
+    setMountNode();
+    // Only rerender if needed
+    if (!disablePortal) {
+      forceUpdate(() => {
```

Comment:
This is not a callback that is executed after a state update was flushed. This is run as a state-update function. These need to be side-effect free. This and the fact that we rely on setTimeout-scheduling the  re-evaluate this approach: What happens if we call `onRender` in an effect that runs if the `mountNode.current` changes?

Reply:
Are you suggesting that whatever is happening in the forceUpdate callback should be in a useEffect so that whenever setMountNode is called, useEffect will run?...something like this:

```
React.useEffect(() => {
        if (onRendered) {
          clearTimeout(renderedTimerRef.current);
          renderedTimerRef.current = setTimeout(onRendered);
        }
      });
  }, [mountNode.current, onRendered]);
```

Reply:
Changed it, Can you please check the latest code, is that what you meant.